src/mem/Allocator.h

```diff
@@ -54,13 +72,62 @@ class Allocator : public Interface
     /// XXX: Misplaced -- not all allocators have a notion of a "chunk". See MemPoolChunked.
     virtual void setChunkSize(size_t) {}
 
+    virtual bool idleTrigger(int shift) const = 0;
+
+    virtual void clean(time_t maxage) = 0;
+
+    /**
+     * Flush temporary counter values into the statistics held in 'meter'.
+     * Performs some CPU intensive calculations which have been deferred for performance.
+     */
+    void flushCounters() {
+        if (count.freed) {
+            getMeter().gb_freed.update(count.freed, objectSize());
+            count.freed = 0;
+        }
+        if (count.allocs) {
+            getMeter().gb_allocated.update(count.allocs, objectSize());
+            count.allocs = 0;
+        }
+        if (count.saved_allocs) {
+            getMeter().gb_saved.update(count.saved_allocs, objectSize());
+            count.saved_allocs = 0;
+        }
+    }
+
     /**
      * \param minSize Minimum size needed to be allocated.
      * \retval n Smallest size divisible by sizeof(void*)
      */
     static size_t RoundedSize(const size_t minSize) { return ((minSize + sizeof(void*) - 1) / sizeof(void*)) * sizeof(void*); }
 
+public:
+
+    /** Temporary counters for performance optimization.
+     *
+     * Counts allocator activity until a flushCounters() gets
+     * triggered to move the values into the 'meter' statistics.
+     *
+     * Specific Optimizations:
+     * \li Objects have fixed-size,
+     *     so all calculations about bytes can be deferred.
+     * \li Objects pooled (if any) can be inferred,
+     *     so saved_free counting is unnecessary.
+     * \li Objects actively in use (if any) can be inferred,
+     *     so no inuse_count management on alloc() or freeOne().
+     */
+    struct FastCounter {
```

Comment:
The above justifications for having FastCounter do not seem to support its existence:

* "Objects have fixed-size, so all calculations about bytes can be deferred." True, but that deference can happen without FastCounter existence AFAICT. In fact, it does happen that way already, at least in one context: `getMeter().gb_allocated.bytes = getMeter().gb_allocated.count * obj_size`. We cannot defer the calculation of the _number_ of allocations (`count.allocs` in the current PR code), but this bullet/justification is not about that; it is about deferring _byte_ calculations.

* "Objects pooled (if any) can be inferred, so saved_free counting is unnecessary." Please rephrase to clarify whether this is about inferring the number of "currently pooled objects" or the number of "saved free() calls (or equivalent)". AFAICT, there is no way to compute saved_free based on allocs, saved_allocs, and freed: If allocs is 2, saved_allocs is 0, and freed is 2, then saved_free could be 2 (because deallocate() added both allocated objects to the pool, after they were allocated,  used, and returned) OR saved_free could could be 0 (because deallocate() freed both returned objects).

* "Objects actively in use (if any) can be inferred, so no inuse_count management on alloc() or freeOne()." True, but we can get the number of objects in use from meter.inuse which is already maintained by deallocate() methods. Thus, we are not saving any CPU cycles for this specific reason.

Please note that I did not check whether we need FastCounter. I would not be very surprised if we do need it. At this point, I am just saying that I do not see how the comment explaining _why_ we need matches code/actual needs.

Reply:
Er, these are not about FastCounter existence. They are justifications for **omitting** the counters that are obviously useful but missing from it.

Reply:
To move forward despite the unresolved disagreements regarding performance, I suggest removing the new FastCounters class from the PR and simply documenting what each counter means, reusing and polishing PR-provided descriptions:

```C++
    /* the three counters below count events since the last flushing */

    /// the number of alloc() calls
    size_t alloc_calls = 0;

    /// the number of alloc() calls that returned a previously allocated object
    size_t saved_allocs = 0;

    /// the number of freeOne() calls
    size_t free_calls = 0; 
```

Replacing FastCounters (including its description) with the above snippet would address this change request.

We could also add a TODO regarding moving these counters to better locations and other improvements, but I doubt we can easily reach consensus regarding those improvements, so I am _not_ suggesting adding any TODOs here.

The data member names above preserve these counter names in the official code. These names are inconsistent and violate current naming guidelines, but we should not bloat this PR with noisy name changes of problematic counters IMO. These names can (and probably should) be changed in a separate PR focused on improving these counters. As far as these counters are concerned, let's limit this PR to moving these counters to the parent class (and adjusting the code accordingly).

Reply:
Done in [52ff68a](https://github.com/squid-cache/squid/pull/1211/commits/52ff68addc2194414a9f6335a8709513dbee799e)